File: src/common/vcs/hg.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Vcs_utils

type error_status = Vcs_utils.error_status =
  | Not_installed of { path: string }
  | Errored of string

let hg ?cwd args =
  let env = [| "NOSCMLOG=1"; "HGPLAIN=1" |] in
  exec ?cwd ~env "hg" args

(** [merge_base a b] returns the hash of the common ancestor of [a] and [b]. *)
let merge_base ?cwd a b =
  let revset = Printf.sprintf "ancestor(%s,%s)" a b in
  match%lwt hg ?cwd ["log"; "-T"; "{node}"; "-r"; revset] with
  | Ok stdout ->
    let result =
      try Ok (String.sub stdout 0 40) with
      | Invalid_argument _ -> Error (Errored "Malformed merge-base")
    in
    Lwt.return result
  | Error _ as err -> Lwt.return err

let files_changed_since ?cwd hash =
  match%lwt hg ?cwd ["status"; "--print0"; "-n"; "--rev"; hash] with
  | Ok stdout -> Lwt.return (Ok (split_null_terminated_lines stdout))
  | Error _ as err -> Lwt.return err

let files_changed_since_mergebase_with ?cwd commit =
  match%lwt merge_base ?cwd "." commit with
  | Error _ as err -> Lwt.return err
  | Ok mergebase ->
    (match%lwt files_changed_since ?cwd mergebase with
    | Ok changes -> Lwt.return (Ok (mergebase, changes))
    | Error _ as err -> Lwt.return err)

File: rust_port/crates/flow_common_vcs/src/hg.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use crate::vcs_utils::Env;
pub use crate::vcs_utils::ErrorStatus;
use crate::vcs_utils::exec;
use crate::vcs_utils::split_null_terminated_lines;

async fn hg(cwd: Option<&str>, args: &[&str]) -> Result<String, ErrorStatus> {
    let env = Env::Extend(vec![
        ("NOSCMLOG".to_string(), "1".to_string()),
        ("HGPLAIN".to_string(), "1".to_string()),
    ]);
    exec(Some(env), cwd, "hg", args).await
}

pub async fn merge_base(cwd: Option<&str>, a: &str, b: &str) -> Result<String, ErrorStatus> {
    let revset = format!("ancestor({},{})", a, b);
    match hg(cwd, &["log", "-T", "{node}", "-r", &revset]).await {
        Ok(stdout) => {
            if stdout.len() >= 40 {
                Ok(stdout[..40].to_string())
            } else {
                Err(ErrorStatus::Errored("Malformed merge-base".to_string()))
            }
        }
        Err(err) => Err(err),
    }
}

pub async fn merge_base_and_timestamp(
    cwd: Option<&str>,
    a: &str,
    b: &str,
) -> Result<(String, i64), ErrorStatus> {
    let revset = format!("ancestor({},{})", a, b);
    match hg(cwd, &["log", "-T", "{node} {date}", "-r", &revset]).await {
        Ok(stdout) => {
            let parts: Vec<&str> = stdout.split(' ').collect();
            match parts.as_slice() {
                [hash, timestamp] => match timestamp.parse::<f64>() {
                    Ok(ts) => Ok((hash.to_string(), ts as i64)),
                    Err(_) => Err(ErrorStatus::Errored("Malformed log response".to_string())),
                },
                _ => Err(ErrorStatus::Errored("Malformed log response".to_string())),
            }
        }
        Err(err) => Err(err),
    }
}

pub async fn files_changed_since(
    cwd: Option<&str>,
    hash: &str,
) -> Result<Vec<String>, ErrorStatus> {
    match hg(cwd, &["status", "--print0", "-n", "--rev", hash]).await {
        Ok(stdout) => Ok(split_null_terminated_lines(&stdout)
            .into_iter()
            .map(|s| s.to_string())
            .collect()),
        Err(err) => Err(err),
    }
}

pub async fn files_changed_since_mergebase_with(
    cwd: Option<&str>,
    commit: &str,
) -> Result<(String, Vec<String>), ErrorStatus> {
    match merge_base(cwd, ".", commit).await {
        Err(err) => Err(err),
        Ok(mergebase) => match files_changed_since(cwd, &mergebase).await {
            Ok(changes) => Ok((mergebase, changes)),
            Err(err) => Err(err),
        },
    }
}
